core/lib/multivm/src/versions/vm_latest/tracers/evm_deploy_tracer.rs

```diff
@@ -0,0 +1,105 @@
+use std::marker::PhantomData;
+
+use zk_evm_1_5_0::{
+    aux_structures::Timestamp,
+    zkevm_opcode_defs::{FatPointer, CALL_IMPLICIT_CALLDATA_FAT_PTR_REGISTER},
+};
+use zksync_contracts::known_codes_contract;
+use zksync_types::{CONTRACT_DEPLOYER_ADDRESS, KNOWN_CODES_STORAGE_ADDRESS};
+use zksync_utils::{bytes_to_be_words, h256_to_u256};
+
+use super::{traits::VmTracer, utils::read_pointer};
+use crate::{
+    interface::{storage::WriteStorage, tracer::TracerExecutionStatus},
+    tracers::dynamic::vm_1_5_0::DynTracer,
+    vm_latest::{
+        utils::hash_evm_bytecode, BootloaderState, HistoryMode, SimpleMemory, ZkSyncVmState,
+    },
+};
+
+/// Tracer responsible for collecting information about EVM deploys and providing those
+/// to the code decommitter.
+#[derive(Debug, Clone)]
+pub(crate) struct EvmDeployTracer<S> {
+    _phantom: PhantomData<S>,
+}
+
+impl<S> EvmDeployTracer<S> {
+    pub(crate) fn new() -> Self {
+        Self {
+            _phantom: PhantomData,
+        }
+    }
+}
+
+impl<S, H: HistoryMode> DynTracer<S, SimpleMemory<H>> for EvmDeployTracer<S> {}
+
+impl<S: WriteStorage, H: HistoryMode> VmTracer<S, H> for EvmDeployTracer<S> {
+    fn finish_cycle(
+        &mut self,
+        state: &mut ZkSyncVmState<S, H>,
+        _bootloader_state: &mut BootloaderState,
+    ) -> TracerExecutionStatus {
+        // We check if ContractDeployer was called with provided evm bytecode.
+        // It is assumed that by that time the user has already paid for its size.
+        // So even if we do not revert the addition of the this bytecode it is not a ddos vector, since
+        // the payment is the same as if the bytecode publication was reverted.
+        let current_callstack = &state.local_state.callstack.current;
+
+        // Here we assume that the only case when PC is 0 at the start of the execution of the contract.
+        let known_code_storage_call = current_callstack.this_address == KNOWN_CODES_STORAGE_ADDRESS
+            && current_callstack.pc == 0
+            && current_callstack.msg_sender == CONTRACT_DEPLOYER_ADDRESS;
```

Comment:
Did this, please check whether the approach is correct.

Reply:
Looks correct. One unrelated thing that I am not sure about is `read_pointer`: it reads region `[start+offset..start+length]`; however in call tracer it doesn't use offset:
```
let packed_abi =
    state.vm_local_state.registers[CALL_IMPLICIT_CALLDATA_FAT_PTR_REGISTER as usize];
assert!(packed_abi.is_pointer);
let far_call_abi = FarCallABI::from_u256(packed_abi.value);
memory.read_unaligned_bytes(
    far_call_abi.memory_quasi_fat_pointer.memory_page as usize,
    far_call_abi.memory_quasi_fat_pointer.start as usize,
    far_call_abi.memory_quasi_fat_pointer.length as usize,
)
```
so it looks like a bug in call_tracer or maybe offset = 0 for some reason. Do you know something about it?

Reply:
Will need to investigate some more, but it could be the case that `offset == 0` for fat pointers to calldata / returndata. E.g., the new VM [narrows the pointer](https://github.com/matter-labs/vm2/blob/a233d44bbe61dc6a758a754c3b78fe4f83e56699/crates/vm2/src/instruction_handlers/far_call.rs#L204) (so that `offset == 0`) in these situations.

Reply:
Looks like the old VM implementation maintains the same invariant: all [calldata](https://github.com/matter-labs/zksync-protocol/blob/a3369809a760c448542d6e6877b95bbabac25d14/crates/zk_evm/src/opcodes/execution/far_call.rs#L328-L335) / [returndata](https://github.com/matter-labs/zksync-protocol/blob/a3369809a760c448542d6e6877b95bbabac25d14/crates/zk_evm/src/opcodes/execution/ret.rs#L86-L93) fat pointers are expected to have `offset == 0`.